### PR TITLE
rework items

### PR DESCRIPTION
**Pull request by Prakkkmak**

### CODE REVIEW

File: code/Items/Inventory.cs [hunk 1]
```diff
@@ -163,7 +131,22 @@ public ItemStack GetItem(int position)
                 throw new IndexOutOfRangeException($"There is only {SlotsCount} slots in the inventory.");
             }
 
-            return items[position];
+            return Slots[position].ItemStack;
+        }
+
+        public Slot GetFirstEmptySlot()
+        {
+            return Slots.FirstOrDefault(slot => slot.IsEmpty());
```
Comment (Nowp) [inline code/Items/Inventory.cs:139]: In the case where no empty slot is found what is returned ?

File: code/Items/Inventory.cs [hunk 2]
```diff
@@ -149,8 +117,8 @@ public ItemStack RemoveItem(int position)
 
             if (!IsSlotEmpty(position))
             {
-                removedItem = items[position];
-                items[position] = null;
+                removedItem = Slots[position].ItemStack;
+                Slots[position].ItemStack = null;
```
Comment (Nowp) [inline code/Items/Inventory.cs:121]: Is it Inventory responsibility to set the ItemStack to null ?

File: code/Items/Slot.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+﻿using Sandbox;
+using ssl.Items.Data;
+
+namespace ssl.Items
+{
+    public class Slot : NetworkComponent
+    {
+        public ItemStack ItemStack { get; set; }
```
Comment (Nowp) [inline code/Items/Slot.cs:8]: Needs to be networked, the class Slot should extend from our workaround NetworkAllTransmitted

File: code/Items/Slot.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        /// <summary>
+        /// Add an itemstack to the slot
+        /// </summary>
+        /// <param name="itemStack">Itemstack to add</param>
+        /// <returns>Remaining itemstack</returns>
+        public ItemStack Add(ItemStack itemStack)
+        {
+            ItemStack ??= new ItemStack(itemStack.Item, 0);
```
Comment (Nowp) [inline code/Items/Slot.cs:27]: what ??= does ?

Comment (GelaPT) [reply code/Items/Slot.cs:27]: I think ??= checks if the item is null and then assigns it a new ItemStack

File: code/Items/Inventory.cs (hunk 1 above)
Comment (Prakkkmak) [reply code/Items/Inventory.cs:139]: null

**Review by Nowp — APPROVED** [review]

**Comment by Prakkkmak** [comment]
[Rework AddItem](https://trello.com/c/MHCNINr7/58-rework-additem)

**Pull request merged**